lib/trivial.nix

```diff
@@ -277,10 +277,17 @@ rec {
      like callPackage expect to be able to query expected arguments.
   */
   setFunctionArgs = f: args:
-    { # TODO: Should we add call-time "type" checking like built in?
-      __functor = self: f;
-      __functionArgs = args;
-    };
+    if builtins.isFunction f then
+      {
+        __functor = self: f;
+        __functionArgs = args;
+      }
+    # Don't wrap with another functor attrset if it's already one
+    else if builtins.isAttrs f && f ? __functor then
+      f // {
+        __functionArgs = args;
+      }
```

Comment:
I’m not sure about the overwriting of existing `__functionArgs` here. What’s the intended semantics?

Reply:
Well the function is called `setFunctionArgs`, so it makes sense that any previous value is dicarded, like `set a . set b == set a` in Haskell land

Reply:
Ah, let’s add it to the documentation then.

But then the question is actually why we might run into a situation where we call set multiple times in the first place (in the calling code).

Reply:
Hehe yes, that's why this function was added in the first place, see your own comment in https://github.com/NixOS/nixpkgs/pull/67809#discussion_r326875446

Reply:
Yeah, but I’m pretty sure the interface is broken if calling `makeOverridable` multiple times overwrites the previous result.

It’s essentially the focus of this PR, so that’s a pretty important nitpick I’d say. :)

Reply:
Ah yeah you actually got something there. The case of a doubly-wrapped functor can never occur in `makeOverridable`!

The only potential candidates are the `setFunctionArgs` in `copyArgs`, which is only ever called with a newly constructed function (no functor possible), or in the `lib.isFunction result` case, but because `isAttrs` is true for functors too, it would have been caught in the previous condition already, therefore `result` can't be a functor either.

Reply:
So it really doesn't matter for this PR, I'll get rid of the `setFunctionArgs` double-wrap-prevention commit. It's just added complexity without any use